docs/proposals/20220210-md-node-label-sync.md

```diff
@@ -0,0 +1,161 @@
+
+Similarly, today the machine controller watches Nodes within the workload cluster, where changes to the Node trigger a reconcile of its corresponding Machine object. This should allow us to capture changes made directly to the labels on the Node (e.g. user accidently deletes a label that falls under CAPI's managed prefix).
+
+#### One-time Apply of Standard Kubernetes Labels
+
+The machine controller will apply the standard kubernetes labels, if specified, to the Node immediately after the Node enters the Ready state (ProviderID is set...).
+
+#### Delay between Node Create and Label Sync 
+
+The Node object is first created when kubeadm joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the LSC kicks in to apply the labels on the Node.
+
+Kubernetes supports both equality and inequality requirements in label selection. In an equality based selection, the user wants to place a workload on node(s) matching a specific label (e.g. Node.Labels contains `my.prefix/foo=bar`). The delay in applying the label on the node, may cause a subsequent delay in the placement of the workload, but this is likely acceptable.
+
+In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.
+
+One way to address this is to use kubelet's `--register-with-taints` flag. We could use that to taint a newly minted Node with say `cluster.x-k8s.io=label-sync-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. When the LSC synchronizes labels, it removes this specific taint.
+
+KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints). We have some options as to how best to “insert” this specific taint: 
+(1) Leave it to the user to specify this exact taint in their KubeadmConfigTemplate manifests
+(2) CABPK inserts it by default in all join configurations. 
+
+Option (1) is the simplest and could be documented as so:
+
+```
+If you utilize inequality based selection for workload placement, to prevent unintended scheduling of pods during the initial node startup phase, it is recommend that you specify the following taint in your KubeadmConfigTemplate:
+`cluster.x-k8s.io=label-sync-pending:NoSchedule`
+```
+
+## Alternatives
+
+### Label propogation from MachineDeployment to Machine
+
+Option #1: Introduce a new field Spec.NodeLifecycle.Labels
+
+Instead of using the top-level metadata, a new field would be introduced:
+
+```
+MachineDeployment.Spec.NodeLifecycle.Labels => MachineSet.Spec.NodeLifecycle.Labels => Machine.Spec.NodeLifecycle.Labels
```

Comment:
I personally prefer this option given that it introduces a clear and explicit boundary between machine labels and node labels.
However, if we go down this path we should consider also impact on ClusterClass and managed topologies

Reply:
I'm fine with both options. But I think I would use `MachineDeployment.spec.template.node.metadata.labels` instead.

But let's get consensus on this before refactoring the proposal back and forth. I think the last time our tendency (@vincepri) was to re-use the existing labels map.

Reply:
Yes, during one of our office hours, the desire was to reuse the existing label propagation.

Reply:
To me `Machine.Spec.NodeLifecycle.Labels` which would also eventually contain drainTimeout and NodeDeletionTimeout is the most intuitive.

Reply:
nit: I think my point is mainly I don't see the point in the "Lifecycle" suffix if labels are part of the node lifecycle, what isn't? :)

Reply:
I guess we need to reach some agreement on this. Is @vincepri going to make the final call?
Whether we use MD.Medata.Label or MD.spec.template.node.metadata.labels, we'll need to support in-place propagation similar to https://github.com/kubernetes-sigs/cluster-api/issues/5880 , since today both scenarios would trigger a rollout, right?

Reply:
> we'll need to support in-place propagation similar to https://github.com/kubernetes-sigs/cluster-api/issues/5880 , since today both scenarios would trigger a rollout, right?

Yes in both cases we need in-place propagation, but only with the new field we're able to avoid a behavioral change (see https://github.com/kubernetes-sigs/cluster-api/pull/6255/files#r881737137).

I'm now leaning towards having a separate field. I personally would prefer `MachineDeployment.spec.template.node.labels` over something with `lifecycle` as `lifecycle` sounds very very generic to me so I'm not sure if it's beneficial to have it in the field name.

I think @vincepri's opinion would be good to hear

Reply:
If we're planning to expand the `node` field in the future, `node.labels` works for me. I'm also fine re-using the usual labels, and only sync the ones with the pre-defined prefix

Reply:
As written above, my tendency would be to use a separate field to make it clear. But I don't want to block if the consensus is to reuse the existing labels field and only sync some of them (the one defined in this proposal) to nodes.

Reply:
Resolving given we had consensus on reusing the existing labels field